src/Compilers/CSharp/Test/Emit3/Semantics/DictionaryExpressionTests.cs

```diff
@@ -1590,35 +1588,107 @@ public MyKeyValuePair(K key, V value)
             var comp = CreateCompilation(sourceA);
             var refA = comp.EmitToImageReference();
 
-            string sourceB = """
+            string sourceB = $$"""
                 using System.Collections.Generic;
                 class Program
                 {
                     static void Main()
                     {
                         var x = new MyKeyValuePair<int, string>(2, "two");
                         var y = new MyKeyValuePair<int, string>[] { new(3, "three") };
-                        F1(x, y);
-                        F2(x, y);
+                        F(x, y).Report();
                     }
-                    static IEnumerable<KeyValuePair<K, V>> F1<K, V>(MyKeyValuePair<K, V> x, IEnumerable<MyKeyValuePair<K, V>> y)
+                    static {{typeName}} F<K, V>(MyKeyValuePair<K, V> x, IEnumerable<MyKeyValuePair<K, V>> y)
                     {
                         return [x, ..y];
                     }
-                    static IDictionary<K, V> F2<K, V>(MyKeyValuePair<K, V> x, IEnumerable<MyKeyValuePair<K, V>> y)
+                }
+                """;
+            var verifier = CompileAndVerify(
+                [sourceB, s_dictionaryExtensions],
+                references: [refA],
+                parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion),
+                expectedOutput: "[2:two, 3:three], ");
+            verifier.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void KeyValuePairConversions_08()
+        {
+            string sourceA = """
+                using System.Collections.Generic;
+                public class MyKeyValuePair<K, V>
+                {
+                    public MyKeyValuePair(K key, V value)
                     {
-                        return [x, ..y];
+                        Key = key;
+                        Value = value;
                     }
+                    public readonly K Key;
+                    public readonly V Value;
+                    public static implicit operator KeyValuePair<K, V>(MyKeyValuePair<K, V> kvp) => new(kvp.Key, kvp.Value);
```

Comment:
Consider including a test for order of evaluation, including a write in the conversion from MyKeyValuePair.